projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py

```diff
@@ -0,0 +1,246 @@
+
+from .kwinner_functions import sampled_kwinners, sampled_kwinners2d
+
+
+def update_temperature(m):
+    """Function used to update SampledKWinner modules temperature. This is typically done
+    during training at the beginning of each epoch.
+
+    Call using :meth:`torch.nn.Module.apply` after each epoch if required
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, SampledKWinnersBase):
+        m.update_temperature()
+
+
+class SampledKWinnersBase(nn.Module, metaclass=abc.ABCMeta):
+    """Base SampledKWinners class.
+
+    :param percent_on:
+      k = percent_on * number of input units will be sampled, with replacement, from
+      categorial softmax distribution over the layer activations
+    :type percent_on: float
+
+    :param k_inference_factor:
+      During inference (training=False) we increase percent_on by this factor.
+      percent_on * k_inference_factor must be strictly less than 1.0, ideally much
+      lower than 1.0
+    :type k_inference_factor: float
+
+    :param temperature:
+      temperature to use when passing activations through softmax. Must be >= 1.0
+    :type temperature: float
+
+    :param temperature_decay_rate:
+      reduce the temperature by this much after each epoch, until it hits 1.0.
+    :type temperature_decay_rate: float
+
+    :param eval_temperature:
+      The temperature to use when passing activations through softmax during evaluation.
+      Must be >= 1.0.
+    :type eval_temperature: float
+    """
+
+    def __init__(
+        self,
+        percent_on,
+        k_inference_factor=1.0,
+        temperature=10.0,
+        temperature_decay_rate=0.01,
+        eval_temperature=1.0,
+    ):
+        super(SampledKWinnersBase, self).__init__()
+        assert 0.0 < percent_on < 1.0
+        assert 0.0 < percent_on * k_inference_factor < 1.0
+        assert eval_temperature >= 1.0
+
+        self.percent_on = percent_on
+        self.percent_on_inference = percent_on * k_inference_factor
+        self.temperature = temperature
+        self.temperature_decay_rate = temperature_decay_rate
+        self.eval_temperature = eval_temperature
+        self.learning_iterations = 0
+        self.n = 0
+        self.k = 0
+        self.k_inference = 0
+
+    def _load_from_state_dict(self, *args, **kwargs):
+        super()._load_from_state_dict(*args, **kwargs)
+
+    def extra_repr(self):
+        return (
+            "n={0},"
+            " percent_on={1},"
+            " temperature={2},"
+            " tempreature_decay_rate={3},"
+            " eval_temperature={4}".format(
+                self.n,
+                self.percent_on,
+                self.temperature,
+                self.temperature_decay_rate,
+                self.eval_temperature
+            )
+        )
+
+    def update_temperature(self):
+        new_temp = self.temperature - self.temperature_decay_rate
```

Comment:
Why subtract the decay rate vs. multiply it? are the precedents in other probabilistic algorithms that you know that do it this way? Or is the another reason?

Reply:
I think it depends on whether we want a linear or exponential decay on the temperature. Idk if there is a standard one way or another

Reply:
Changed it to exponential decay since that does seem more appropriate.

Reply:
Sounds good. We used exponential decay for the boost strength, so I think that's more consistent. Nonetheless, it's an open question. I'll add it to the list of things to consider for Alex.